src/admin/__tests__/deleteTopicRecords.spec.js

```diff
@@ -0,0 +1,269 @@
+        const value = secureRandom()
+        return { key: `key-${i}`, value: `value-${value}` }
+      })
+
+    await producer.send({ acks: 1, topic: topicName, messages })
+
+    groupId = `consumer-group-id-${secureRandom()}`
+    consumer = createConsumer({
+      cluster,
+      groupId,
+      maxWaitTimeInMs: 100,
+      logger,
+    })
+    await consumer.subscribe({ topic: topicName, fromBeginning: true })
+
+    // validate the modulus partitioner allocates 20 messages 13:7
+    expect(
+      await cluster.fetchTopicsOffset([
+        {
+          topic: topicName,
+          partitions: [{ partition: 0 }, { partition: 1 }],
+          fromBeginning: false,
+        },
+      ])
+    ).toEqual([
+      {
+        topic: topicName,
+        partitions: expect.arrayContaining([
+          { partition: 0, offset: '13' },
+          { partition: 1, offset: '7' },
+        ]),
+      },
+    ])
+  })
+
+  afterEach(async () => {
+    producer && (await producer.disconnect())
+    admin && (await admin.disconnect())
+    consumer && (await consumer.disconnect())
+    jest.resetAllMocks()
+  })
+
+  afterAll(jest.restoreAllMocks)
+
+  test('throws an error if the topic name is invalid', async () => {
+    await expect(admin.deleteTopicRecords({ topic: null })).rejects.toHaveProperty(
+      'message',
+      'Invalid topic null'
+    )
+
+    await expect(admin.deleteTopicRecords({ topic: ['topic-in-an-array'] })).rejects.toHaveProperty(
+      'message',
+      'Invalid topic topic-in-an-array'
+    )
+  })
+
+  test('throws an error if the partitions array is invalid', async () => {
+    await expect(
+      admin.deleteTopicRecords({ topic: topicName, partitions: [] })
+    ).rejects.toHaveProperty('message', 'Invalid partitions')
+  })
+
+  test('removes deleted offsets from the selected partition', async () => {
+    recordsToDelete = [{ partition: 0, offset: '7' }]
+    await admin.deleteTopicRecords({ topic: topicName, partitions: recordsToDelete })
+
+    expect(
+      await cluster.fetchTopicsOffset([
+        {
+          topic: topicName,
+          partitions: [{ partition: 0 }, { partition: 1 }],
+          fromBeginning: true,
+        },
+      ])
+    ).toEqual([
+      {
+        topic: topicName,
+        partitions: expect.arrayContaining([
+          { partition: 0, offset: '7' },
+          { partition: 1, offset: '0' },
+        ]),
+      },
+    ])
+  })
+
+  test('non-deleted messages are successfully consumed', async () => {
+    recordsToDelete = [{ partition: 0, offset: '7' }]
+    const messagesPartition0 = []
+    const messagesPartition1 = []
+    await admin.deleteTopicRecords({ topic: topicName, partitions: recordsToDelete })
+    consumer.run({
+      eachMessage: async event => {
+        if (event.partition === 0) messagesPartition0.push(event)
+        if (event.partition === 1) messagesPartition1.push(event)
+      },
+    })
+    // wait for the 2 partition batches to get processed
```

Comment:
cool, just updated to use `waitForMessages`. I had seen it, and initially I had leant towards waiting on the batch events as a way to simulate "wait for _every_ available message to get consumed" rather than waiting for a set number, because there's a bit of circularity in waiting for 13 messages to get consumed, then testing that each partition has the correct split of these 13 messages? (you will always have 13 because that's what you waited for)

With that in mind I've added an extra assertion to check that the first message consumed on partition 0 is offset 7, and the last message is offset 12, so that it's clear that the first 7 were skipped, hopefully that covers everything while not too clunky :)